picard/album.py

```diff
@@ -467,54 +468,53 @@ def _remove_file(self, track, file):
         file.metadata_images_changed.disconnect(self.update_metadata_images)
         remove_metadata_images(self, [file])
 
-    def match_files(self, files, use_recordingid=True):
+    def _match_files(self, files, recordingid=None, threshold=0):
         """Match files to tracks on this album, based on metadata similarity or recordingid."""
+        moves = []
+        tracks_cache = defaultdict(lambda: None)
         for file in list(files):
             if file.state == File.REMOVED:
                 continue
-            matches = []
-            recordingid = file.metadata['musicbrainz_recordingid']
-            if use_recordingid and mbid_validate(recordingid):
-                matches = self._get_recordingid_matches(file, recordingid)
-            if not matches:
-                for track in self.tracks:
-                    sim = track.metadata.compare(file.orig_metadata)
-                    if sim >= config.setting['track_matching_threshold']:
-                        matches.append((sim, track))
-            if matches:
-                matches.sort(key=itemgetter(0), reverse=True)
-                file.move(matches[0][1])
-            else:
-                file.move(self.unmatched_files)
-
-    def match_file(self, file, recordingid=None):
-        """Match the file on a track on this album, based on recordingid or metadata similarity."""
-        if file.state == File.REMOVED:
-            return
-        if recordingid is not None:
-            matches = self._get_recordingid_matches(file, recordingid)
-            if matches:
-                matches.sort(key=itemgetter(0), reverse=True)
-                file.move(matches[0][1])
-                return
-        self.match_files([file], use_recordingid=False)
-
-    def _get_recordingid_matches(self, file, recordingid):
-        matches = []
-        tracknumber = file.metadata['tracknumber']
-        discnumber = file.metadata['discnumber']
-        for track in self.tracks:
-            tm = track.orig_metadata
-            if recordingid == tm['musicbrainz_recordingid']:
-                if tracknumber == tm['tracknumber']:
-                    if discnumber == tm['discnumber']:
-                        matches.append((4.0, track))
-                        break
-                    else:
-                        matches.append((3.0, track))
-                else:
-                    matches.append((2.0, track))
-        return matches
+            # if we have a recordingid to match against, use that in priority
+            recid = recordingid or file.metadata['musicbrainz_recordingid']
+            if recid and mbid_validate(recid):
+                if not tracks_cache:
```

Comment:
> This cache could be built outside the loop. tracks_cache is only empty on the first iteration.

Yes, it could, but if we are in the case there are no recordingids to match against, the cache would be built for nothing. If `recordingid` parameter is set and is a valid mbid, we are certain this cache will be used, but if not, it depends on `file.metadata['musicbrainz_recordingid']` and it can only be checked within the loop.
So i don't think we can actually really improve things there, that's just a simple test and overall we reduced the number of them with those changes.
I could move the code outside the loop to a function, but it would add a function one-time function call, so it'd be "cosmetic" only (and less efficient).

The second part (similarity matching) is very slow compared to this part.

Reply:
> I could move the code outside the loop to a function, but it would add a function one-time function call, so it'd be "cosmetic" only (and less efficient).

I did that, it makes code easier to read, and the cost of one function called once isn't significant

Reply:
Yes, I think this is better this way